# Project 9 -- Niharika Rai

**TA Help:** (for instance) John Smith, Alice Jones, etc., list names of any TAs who helped you

- For example: Help with figuring out how to write a function (describe the tasks that they helped you with)

**Collaboration:** My Friend in CS, My Uncle, Another Student, etc., list names of any other people who helped you

(describe the tasks that they helped you with)
- For example: helped figuring out how to load the dataset.
- Another example: helped debug error with my plot.

**Internet Resources:** Stack Exchange, Stack Overflow, etc.

(describe any information that you learned from internet resources, including the URLs)
- data frames in Pandas versus R from StackOverflow  https://stackoverflow.com/questions/8991709/why-were-pandas-merges-in-python-faster-than-data-table-merges-in-r-in-2012

**ChatGPT, Gemini, Claude, etc:** Any language models or generative AI chatbots that helped you.

(if you used any such tools, please tell us here)
- For example:  I asked ChatGPT how to define a new data frames
- Another example:  Gemini told me how to make a function for sorting my data

- ***Link to AI Chat History***: Here's the link: https://chatgpt.com/share/69005ff3-efa8-800d-8626-cc4bbd6120ec. Please share a link to your chat if you used AI (ex. ChatGPT Shared Links)
**OVERALL MESSAGE:** Any time that you used anything except your brain to solve the questions in these projects, you need to disclose such resources at the start of the project, with details about your usage of the tools.

**YOUR OWN WORK:** Even when you utilize other resources, do NOT just copy and paste.  Write all explanations in your own words, using several sentences in English, which are understandable and which you wrote (and did not just copy and paste).

## Question 1

In [1]:
%sql sqlite:///./lahman.db

In [2]:
%%sql
SELECT * FROM Teams LIMIT 5;

 * sqlite:///./lahman.db
Done.


yearID,lgID,teamID,franchID,divID,teamRank,G,Ghome,W,L,DivWin,WCWin,LgWin,WSWin,R,AB,H,2B,3B,HR,BB,SO,SB,CS,HBP,SF,RA,ER,ERA,CG,SHO,SV,IPouts,HA,HRA,BBA,SOA,E,DP,FP,name,park,attendance,BPF,PPF,teamIDBR,teamIDlahman45,teamIDretro
1871,NA,BS1,BNA,,3,31,,20,10,,,N,,401,1372,426,70,37,3,60,19,73,16,,,303,109,3.55,22,1,3,828,367,2,42,23,243,24,0.834,Boston Red Stockings,South End Grounds I,,103,98,BOS,BS1,BS1
1871,NA,CH1,CNA,,2,28,,19,9,,,N,,302,1196,323,52,21,10,60,22,69,21,,,241,77,2.76,25,0,1,753,308,6,28,22,229,16,0.829,Chicago White Stockings,Union Base-Ball Grounds,,104,102,CHI,CH1,CH1
1871,NA,CL1,CFC,,8,29,,10,19,,,N,,249,1186,328,35,40,7,26,25,18,8,,,341,116,4.11,23,0,0,762,346,13,53,34,234,15,0.818,Cleveland Forest Citys,National Association Grounds,,96,100,CLE,CL1,CL1
1871,NA,FW1,KEK,,7,19,,7,12,,,N,,137,746,178,19,8,2,33,9,16,4,,,243,97,5.17,19,1,0,507,261,5,21,17,163,8,0.803,Fort Wayne Kekiongas,Hamilton Field,,101,107,KEK,FW1,FW1
1871,NA,NY2,NNA,,5,33,,16,17,,,N,,302,1404,403,43,21,1,33,15,46,15,,,313,121,3.72,32,1,0,879,373,7,42,22,235,14,0.84,New York Mutuals,Union Grounds (Brooklyn),,90,88,NYU,NY2,NY2


In [3]:
%%sql
SELECT * FROM People LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,birthYear,birthMonth,birthDay,birthCity,birthCountry,birthState,deathYear,deathMonth,deathDay,deathCountry,deathState,deathCity,nameFirst,nameLast,nameGiven,weight,height,bats,throws,debut,bbrefID,finalGame,retroID
aardsda01,1981,12,27,Denver,USA,CO,,,,,,,David,Aardsma,David Allan,215,75,R,R,2004-04-06,aardsda01,2015-08-23,aardd001
aaronha01,1934,2,5,Mobile,USA,AL,2021,1,22,USA,GA,Atlanta,Hank,Aaron,Henry Louis,180,72,R,R,1954-04-13,aaronha01,1976-10-03,aaroh101
aaronto01,1939,8,5,Mobile,USA,AL,1984,8,16,USA,GA,Atlanta,Tommie,Aaron,Tommie Lee,190,75,R,R,1962-04-10,aaronto01,1971-09-26,aarot101
aasedo01,1954,9,8,Orange,USA,CA,,,,,,,Don,Aase,Donald William,190,75,R,R,1977-07-26,aasedo01,1990-10-03,aased001
abadan01,1972,8,25,Palm Beach,USA,FL,,,,,,,Andy,Abad,Fausto Andres,184,73,L,L,2001-09-10,abadan01,2006-04-13,abada001


In [4]:
%%sql 
CREATE TRIGGER IF NOT EXISTS update_yearID
AFTER INSERT ON Teams
FOR EACH ROW
BEGIN
    UPDATE Teams
    SET yearID = strftime('%Y', 'now')
    WHERE teamID = NEW.teamID;
END;

 * sqlite:///./lahman.db
Done.


[]

In [5]:
%%sql 
INSERT INTO Teams (teamID, name, yearID)
VALUES ('team1', 'Test Team', 0);

 * sqlite:///./lahman.db
1 rows affected.


[]

In [6]:
%%sql 
SELECT * FROM Teams WHERE teamID = 'team1';

 * sqlite:///./lahman.db
Done.


yearID,lgID,teamID,franchID,divID,teamRank,G,Ghome,W,L,DivWin,WCWin,LgWin,WSWin,R,AB,H,2B,3B,HR,BB,SO,SB,CS,HBP,SF,RA,ER,ERA,CG,SHO,SV,IPouts,HA,HRA,BBA,SOA,E,DP,FP,name,park,attendance,BPF,PPF,teamIDBR,teamIDlahman45,teamIDretro
2025,None,team1,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,Test Team,None,None,None,None,None,None,None
2025,None,team1,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,Test Team,None,None,None,None,None,None,None
2025,None,team1,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,Test Team,None,None,None,None,None,None,None
2025,None,team1,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,None,Test Team,None,None,None,None,None,None,None


1.1 I used the query given in the question. 
1.2 I used the query given in the question. 
1.3 I used the query given in the question. My output shows that the yearID has been automatically updated to the current year, 2025, even though I originally inserted a value of 0.   

## Question 2

In [7]:
%%sql 
CREATE TABLE DeathRecords (
    playerID TEXT PRIMARY KEY,
    yearID INTEGER NOT NULL,
    monthID INTEGER NOT NULL,
    dayID INTEGER NOT NULL
);

 * sqlite:///./lahman.db
Done.


[]

In [8]:
%%sql 
CREATE TRIGGER IF NOT EXISTS update_deathYearMonthDay
AFTER INSERT ON DeathRecords
FOR EACH ROW
BEGIN
    UPDATE People
    SET deathYear = NEW.yearID,
        deathMonth = NEW.monthID,
        deathDay = NEW.dayID
    WHERE People.playerID = NEW.playerID;
END;

 * sqlite:///./lahman.db
Done.


[]

In [9]:
%%sql 
INSERT INTO DeathRecords (playerID, yearID, monthID, dayID)
VALUES ('abreubo01', 2025, 1, 1);

 * sqlite:///./lahman.db
1 rows affected.


[]

In [10]:
%%sql 
SELECT * FROM People WHERE playerID = 'abreubo01';

 * sqlite:///./lahman.db
Done.


playerID,birthYear,birthMonth,birthDay,birthCity,birthCountry,birthState,deathYear,deathMonth,deathDay,deathCountry,deathState,deathCity,nameFirst,nameLast,nameGiven,weight,height,bats,throws,debut,bbrefID,finalGame,retroID
abreubo01,1974,3,11,Maracay,Venezuela,Aragua,2025,1,1,,,,Bobby,Abreu,Bob Kelly,220,72,L,R,1996-09-01,abreubo01,2014-09-28,abreb001


In [11]:
%%sql 
INSERT INTO DeathRecords (playerID, yearID, monthID, dayID)
VALUES
('pintore01', 2025, 2, 1),
('hundlni01', 2025, 3, 1),
('detwiro01', 2025, 4, 1);

 * sqlite:///./lahman.db
3 rows affected.


[]

In [12]:
%%sql 
SELECT playerID, deathYear, deathMonth, deathDay FROM People
WHERE playerID IN ('pintore01', 'hundlni01', 'detwiro01');

 * sqlite:///./lahman.db
Done.


playerID,deathYear,deathMonth,deathDay
detwiro01,2025,4,1
hundlni01,2025,3,1
pintore01,2025,2,1


2.1 I used the 'CREATE TABLE' syntax from project 8 and followed the question's instructions when assigning the primary key and column data types when creating the table. 
2.2 I first created my own trigger using the syntax in the sample query of this sub-question. However, I ran into syntax errors and used ChatGPT to debug my code. After I corrected my code, it worked. 
2.3 I used the sample query from the question to verify this and it works. 
2.4 I used the sample query from the question to verify this and it works.

## Question 3

In [13]:
%%sql 
CREATE VIEW IF NOT EXISTS TallPlayers AS
SELECT playerID, nameFirst, nameLast, height, weight
FROM People
WHERE height > 72;

 * sqlite:///./lahman.db
Done.


[]

In [14]:
%%sql 
SELECT * FROM TallPlayers LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,nameFirst,nameLast,height,weight
aardsda01,David,Aardsma,75,215
aaronto01,Tommie,Aaron,75,190
aasedo01,Don,Aase,75,190
abadan01,Andy,Abad,73,184
abadfe01,Fernando,Abad,74,235


In [15]:
%%sql 
CREATE VIEW IF NOT EXISTS TallHeavyPlayers AS
SELECT playerID, nameFirst, nameLast, height, weight
FROM People
WHERE height > 72 AND weight IS NOT NULL AND birthYear = 1994
ORDER BY weight DESC;

 * sqlite:///./lahman.db
Done.


[]

In [16]:
%%sql 
SELECT * FROM TallHeavyPlayers LIMIT 5;

 * sqlite:///./lahman.db
Done.


playerID,nameFirst,nameLast,height,weight
castidi01,Diego,Castillo,75,268
perdoan01,Angel,Perdomo,80,265
ponceco01,Cody,Ponce,78,255
quezajo01,Johan,Quezada,81,255
wellsty01,Tyler,Wells,80,255


In [17]:
%%sql
SELECT * FROM TallHeavyPlayers LIMIT 7;

 * sqlite:///./lahman.db
Done.


playerID,nameFirst,nameLast,height,weight
castidi01,Diego,Castillo,75,268
perdoan01,Angel,Perdomo,80,265
ponceco01,Cody,Ponce,78,255
quezajo01,Johan,Quezada,81,255
wellsty01,Tyler,Wells,80,255
alonspe01,Pete,Alonso,74,253
bednada01,David,Bednar,73,250


3.1 I used the sample query in the question as a reference and modified it get the required output. 
3.2 I used the sample query in the question to verify if all the players were taller than 6 feet. 
3.3 I used the query created in 3.1 but changed the view name and added more conditions to get the required output. 
3.4 I used the sample query in the question to get the top 5 heaviest players, born in 1994 and are over 6 feet tall.   
3.5 I used the sample query in the question to get the top 7 heaviest players, born in 1994 and are over 6 feet tall.   

## Question 4

In [36]:
%config SqlMagic.autocommit = False

In [37]:
%%sql
SAVEPOINT insert_deathrecords;

INSERT INTO DeathRecords VALUES ('bandyje01', 2025, 5, 1);
INSERT INTO DeathRecords VALUES ('meyeral01', 2025, 6, 1);
INSERT INTO DeathRecords VALUES ('perdoan01', 2025, 7, 1);

RELEASE insert_deathrecords;
COMMIT;

 * sqlite:///./lahman.db
Done.
(sqlite3.IntegrityError) UNIQUE constraint failed: DeathRecords.playerID
[SQL: INSERT INTO DeathRecords VALUES ('bandyje01', 2025, 5, 1);]
(Background on this error at: https://sqlalche.me/e/20/gkpj)


In [38]:
%%sql 
SELECT * FROM DeathRecords;

 * sqlite:///./lahman.db
Done.


playerID,yearID,monthID,dayID
abreubo01,2025,1,1
pintore01,2025,2,1
hundlni01,2025,3,1
detwiro01,2025,4,1
bandyje01,2025,5,1
meyeral01,2025,6,1
perdoan01,2025,7,1


In [39]:
%%sql 
BEGIN;
SAVEPOINT deathrecords_insert2;
INSERT OR IGNORE INTO DeathRecords VALUES ('castidi01', 2025, 8, 1);
INSERT OR IGNORE INTO DeathRecords VALUES ('quezajo01', 2025, 9, NULL);

ROLLBACK TO deathrecords_insert2;
COMMIT;

 * sqlite:///./lahman.db
Done.
Done.
1 rows affected.
0 rows affected.
Done.
Done.


[]

In [40]:
%%sql 
SELECT * FROM DeathRecords WHERE playerID = 'castidi01';

 * sqlite:///./lahman.db
Done.


playerID,yearID,monthID,dayID


4.1 I used the sample query in the question to achieve this. 
4.2 I used the sample query in the question to verify that 3 new rows have been added.  
4.3 I used the sample query in the question to achieve this. 
4.4 I used the sample query in the question to verify that no new rows were inserted.  

## Question 5

In [76]:
%%sql 
CREATE TRIGGER IF NOT EXISTS update_myplayers_death
AFTER INSERT ON DeathRecords
FOR EACH ROW
BEGIN
    UPDATE People
    SET deathYear = NEW.yearID,
        deathMonth = NEW.monthID,
        deathDay = NEW.dayID
    WHERE playerID = NEW.playerID;
END;

 * sqlite:///./lahman.db
Done.


[]

In [77]:
%%sql 
CREATE VIEW IF NOT EXISTS DeceasedPlayers AS
SELECT playerID, nameFirst, nameLast, deathYear, deathMonth, deathDay
FROM People
WHERE deathYear IS NOT NULL;

 * sqlite:///./lahman.db
Done.


[]

In [78]:
%%sql
SAVEPOINT insert_deathrecords;

INSERT INTO DeathRecords VALUES ('johndoe01', 2025, 10, 1);
INSERT INTO DeathRecords VALUES ('janedoe01', 2025, 11, 1);
INSERT INTO DeathRecords VALUES ('foobar01', 2025, 12, 1);

RELEASE insert_deathrecords;
COMMIT;

 * sqlite:///./lahman.db
Done.
(sqlite3.IntegrityError) UNIQUE constraint failed: DeathRecords.playerID
[SQL: INSERT INTO DeathRecords VALUES ('johndoe01', 2025, 10, 1);]
(Background on this error at: https://sqlalche.me/e/20/gkpj)


In [79]:
%%sql 
SELECT * FROM DeathRecords WHERE playerID IN ('johndoe01', 'janedoe01', 'foobar01');

 * sqlite:///./lahman.db
Done.


playerID,yearID,monthID,dayID
foobar01,2025,12,1
janedoe01,2025,11,1
johndoe01,2025,10,1


In [82]:
%%sql 
SELECT * FROM DeceasedPlayers WHERE playerID IN ('johndoe01', 'janedoe01', 'foobar01');

 * sqlite:///./lahman.db
Done.


playerID,nameFirst,nameLast,deathYear,deathMonth,deathDay


5.1 I used the sample query in the question to achieve this.
5.2 I used the sample query in the question to achieve this.
5.3 I used the sample query in the question but modified it based on the different row information that I was supposed to insert. 
5.4 I used the sample query in the question to achieve this.
5.5 I used the sample query in the question to achieve this. However, the view table does not show the 3 rows added. It shows that the columns are present but they are no rows added.  

## Pledge

By submitting this work I hereby pledge that this is my own, personal work. I've acknowledged in the designated place at the top of this file all sources that I used to complete said work, including but not limited to: online resources, books, and electronic communications. I've noted all collaboration with fellow students and/or TA's. I did not copy or plagiarize another's work.

> As a Boilermaker pursuing academic excellence, I pledge to be honest and true in all that I do. Accountable together – We are Purdue.

https://www.purdue.edu/odos/osrr/honor-pledge/
